src/main/java/de/bonndan/nivio/input/ItemRelationProcessor.java

```diff
@@ -18,69 +18,129 @@ protected ItemRelationProcessor(ProcessLog processLog) {
     }
 
     @Override
-    public void process(LandscapeDescription input, Landscape landscape) {
-        input.getItemDescriptions().all().forEach(serviceDescription -> {
-            Item origin = landscape.getItems().pick(serviceDescription);
-            if (!input.isPartial()) {
-                processLog.debug(String.format("Clearing relations of %s", origin));
-                origin.getRelations().clear(); //delete all relations on full update
-            }
-        });
+    public ProcessingChangelog process(LandscapeDescription input, Landscape landscape) {
+
+        ProcessingChangelog changelog = new ProcessingChangelog();
+        List<Relation> processed = new ArrayList<>();
 
         input.getItemDescriptions().all().forEach(itemDescription -> {
             Item origin = landscape.getItems().pick(itemDescription);
+            List<Relation> affected = new ArrayList<>();
 
             for (RelationDescription relationDescription : itemDescription.getRelations()) {
-                Optional<Relation> update = update(relationDescription, landscape, origin);
-                update.ifPresent(relation -> {
-                    origin.getRelations().remove(relation);
-                    origin.getRelations().add(relation);
-                    if (relation.getSource() == origin) {
-                        relation.getTarget().getRelations().remove(relation);
-                        relation.getTarget().getRelations().add(relation);
-                    } else {
-                        relation.getSource().getRelations().remove(relation);
-                        relation.getSource().getRelations().add(relation);
+
+                if (!isValid(relationDescription, landscape)) {
+                    continue;
+                }
+
+                Optional<Relation> existing = getExisting(relationDescription, landscape, origin);
+                if (existing.isPresent()) {
```

Comment:
I think we can find better names for `existing` and `getExisting()` 😸 - `if(existing.isPresent())` should be present by definition 😄 

How about just `getRelation()`? If it is present, it is existent.

Reply:
I understand why you chose `existing`, below we have `created`, but still it reads strangely like it is now :)

Reply:
changed to "current"